### PR TITLE
Add WithTransaction function

### PR DESCRIPTION
**Pull request by levon80999**

### CODE REVIEW

File: src/Jenssegers/Mongodb/Concerns/TransactionManager.php [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+     * create a session and start a transaction in session
+     *
+     * In version 4.0, MongoDB supports multi-document transactions on replica sets.
+     * In version 4.2, MongoDB introduces distributed transactions, which adds support for multi-document transactions on sharded clusters and incorporates the existing support for multi-document transactions on replica sets.
+     * To use transactions on MongoDB 4.2 deployments(replica sets and sharded clusters), clients must use MongoDB drivers updated for MongoDB 4.2.
+     *
+     * @see https://docs.mongodb.com/manual/core/transactions/
+     * @return void
+     */
+    public function beginTransaction(?array $options = [])
```
Comment (alcaeus) [inline src/Jenssegers/Mongodb/Concerns/TransactionManager.php]: The options parameter doesn't need to be nullable; the important part is having a default value so the method is compatible with the parent interface.

```suggestion
    public function beginTransaction(array $options = [])
```

File: src/Jenssegers/Mongodb/Concerns/TransactionManager.php [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+    {
+        $session = $this->getSession();
+
+        if (!$session) {
+            $this->session_key = uniqid();
+            $session = $this->connection->startSession();
+            $this->sessions[$this->session_key] = $session;
```
Comment (alcaeus) [inline src/Jenssegers/Mongodb/Concerns/TransactionManager.php]: Since we're only managing a single session, we can do away with the array and store the single session directly. This also makes the session key unnecessary.

File: src/Jenssegers/Mongodb/Connection.php [hunk 3]
```diff
@@ -65,6 +68,17 @@ public function __construct(array $config)
         $this->useDefaultQueryGrammar();
     }
 
+    /**
+     * Dynamically pass methods to the connection.
+     * @param string $method
+     * @param array $parameters
+     * @return mixed
+     */
+    public function __call($method, $parameters)
```
Comment (alcaeus) [inline src/Jenssegers/Mongodb/Connection.php]: Is this method necessary for transactions? If so, can you explain?

File: src/Jenssegers/Mongodb/Connection.php [hunk 4]
```diff
@@ -93,7 +107,7 @@ public function table($table, $as = null)
      * @param string $name
      * @return Collection
      */
-    public function getCollection($name)
+    public function getCollection(string $name)
```
Comment (alcaeus) [inline src/Jenssegers/Mongodb/Connection.php]: I believe changes like this should be extracted to a separate PR to avoid side-effects.

File: src/Jenssegers/Mongodb/Connection.php [hunk 5]
```diff
@@ -117,17 +131,17 @@ public function getMongoDB()
 
     /**
      * return MongoDB object.
-     * @return \MongoDB\Client
+     * @return Client
      */
-    public function getMongoClient()
+    public function getMongoClient(): Client
```
Comment (alcaeus) [inline src/Jenssegers/Mongodb/Connection.php]: This is a BC break for anyone extending this connection class.

File: src/Jenssegers/Mongodb/Connection.php [hunk 6]
```diff
@@ -117,17 +131,17 @@ public function getMongoDB()
...
     {
         return $this->connection;
     }
 
     /**
      * {@inheritdoc}
      */
-    public function getDatabaseName()
+    public function getDatabaseName(): string
```
Comment (alcaeus) [inline src/Jenssegers/Mongodb/Connection.php]: This is a BC break for anyone extending this connection class.

File: tests/TransactionTest.php [hunk 7]
```diff
@@ -5,9 +5,9 @@
 
 class TransactionTest extends TestCase
 {
-    protected $insertData = ['name' => 'klinson', 'age' => 20, 'title' => 'admin'];
-    protected $originData = ['name' => 'users', 'age' => 20, 'title' => 'user'];
-    protected $connection = 'mongodb_repl';
+    protected array $insertData = ['name' => 'klinson', 'age' => 20, 'title' => 'admin'];
```
Comment (alcaeus) [inline tests/TransactionTest.php]: I'm not sure what PHP version the library requires, as it's not noted in composer.json, but considering typed properties require PHP 7.4 we may want to undo this change.

**Review by alcaeus — CHANGES_REQUESTED** [review]
Some initial comments. @levon80999 can you reset your master branch to the current state in https://github.com/jenssegers/laravel-mongodb so we can review the PR as you'd create it in the Laravel repository? That way we could leave a comprehensive review.

File: src/Jenssegers/Mongodb/Concerns/TransactionManager.php (hunk 1 above)
Comment (levon80999) [reply src/Jenssegers/Mongodb/Concerns/TransactionManager.php]: Yes, but startTransaction declaration has nullable argument that is why  I decide to define it as nullable 
`final public function startTransaction(?array $options = []) {}`

File: src/Jenssegers/Mongodb/Connection.php (hunk 3 above)
Comment (levon80999) [reply src/Jenssegers/Mongodb/Connection.php]: I just move this function to top after __construct method

File: src/Jenssegers/Mongodb/Concerns/TransactionManager.php (hunk 1 above)
Comment (alcaeus) [reply src/Jenssegers/Mongodb/Concerns/TransactionManager.php]: That is a remnant of the current API and will be changed in 2.0, so I'd prefer to pre-empt that change here

File: tests/TransactionTest.php (hunk 7 above)
Comment (jmikola) [reply tests/TransactionTest.php]: This package depends version `^9.0` of various Laravel libraries, which require PHP 8+. See: https://github.com/illuminate/database/blob/v9.0.0/composer.json#L18

@levon80999: Rather than simply revert the change, it would have been preferable to research the actual requirement. There's a difference between @alcaeus _suggesting_ that "we may want to undo this change" and directly asking you to revert it.

A separate follow-up task for this PR (before merging) could be determining the minimum PHP requirement based on the features used within and add a language requirement to `composer.json`. Even if the package might indirectly require PHP 8+ through its Laravel deps, there's no reason we can't accurately note what version the package itself requires based on its own sources.

**Comment by levon80999** [comment]
https://github.com/levon80999/laravel-mongodb/pull/3
